### ISSUE TITLE
[NuGet] Automatically adding the MonoGame project file elements

### ISSUE THREAD

**Issue by mattleibow** [MonoGame/MonoGame#4316]
Hi guys.

I was just wondering if a PR for adding the `<MonoGamePlatform>` and `<Import>` elements to the project files automatically is something that I could do? Each time I install MonoGame into a new project, I have to manually edit the files and add it. Not too hard, but very annoying. We can do this automatically without any strange scripts and things. 

Basically, according to the [docs](http://docs.nuget.org/create/Creating-and-Publishing-a-Package#import-msbuild-targets-and-props-files-into-project), we can add a `.props` file that adds the `<MonoGamePlatform>` element based on the NuGet installed. As each NuGet has a single platform, there will be no confusion. Then, we just add the `.targets` file to the NuGet and it will automatically be referenced.

I have done this with several projects, and is very useful. It works in Visual Studio, Xamarin Studio/MonoDevelop. If anything fails, such as some random IDE, it can still be added manually.

**Comment by tomspilman** [comment]
I assume this is just for when someone adds MonoGame to an existing project via NuGet correct?

@DDReaper @dellis1972 - You guys understand how this works?  Thoughts?

**Comment by DDReaper** [comment]
I toyed with using .Targets and .Props early on with the NuGet development, however I dropped it as some platforms (you know which) didn't like them very much, so I dropped them.
Feel free to work up a new PR and add them to a test package but ensure you test them on all platforms.  Things have moved along so they work better now and it would be a good solution for blank projects.

Ideally though, you should be creating new projects using the templates and then replacing your references with the NuGet's to keep them up to date, except on those platforms that don't have templates.

**Comment by dellis1972** [comment]
.targets should work now on Mac not sure about .props though.. The problem I have with .props is visibility. the current <MonoGamePlatform> settings are put in the first PropertyGroup and are quite visible to the user and easy to change. A .props file is hidden away in the nuget packages folder somewhere so its not quite as obvious..

A ps1 script would do the trick to insert the required stuff in the right place in the csproj, but that will only work on Windows.. But that possibly isn't a problem because the XS Addin does that itself now so Mac is already taken care of as far as the settings and imports are concerned.

generally people should be using the templates if possible.. Perhaps we just need to update the READme on the nuget packages to include instructions..

**Comment by mattleibow** [comment]
I noticed in #4152 that there seems to be more stuff to add:
```
<MonoGameInstallDirectory Condition="'$(OS)' != 'Unix' ">$(MSBuildProgramFiles32)</MonoGameInstallDirectory>
<MonoGameInstallDirectory Condition="'$(OS)' == 'Unix' ">$(MSBuildExtensionsPath)</MonoGameInstallDirectory>
```

This is getting a bit much, as it is just the same. Technically, this isn't a real problem as we can always override the values in the .csproj. If the `<MonoGamePlatform>` element is in the .props, then it is overridable in the first section as well. This also prevents any breakages with updates and all.

**Comment by mattleibow** [comment]
This works with #4152 and #3985

**Comment by dellis1972** [comment]
@mattleibow the issue we have is this..

Nuget packages can add those properties via .props etc which is great. 
Our templates in VS DONT use Nuget packages they use the locally installed assemblies, so we need to have those properties in the main .csproj or have multiple .props files for each platform installed along side the assemblies. If just feels kinda messy. 

One solution would be to have the templates use Nuget Packages but allot of people don't want to do that since at least half of our user base use the installer rather than nuget and it limits offline usage as you will need to install the Nuget packages at least once to get them in the cache.
So how do we resolve this? We could just say Nuget is different I guess and leave the VS templates as they are perhaps..

FYI Xamarin Studio has to use Nugets for its templates because XS is really bad a resolving assemblies, its just something we had to live with but we install the nugets locally as part of the Addin installation. 

I'm all for having  Pipeline Tool Nuget as well CocosSharp does that (would you include the GUI tool in that?) the problem with that is the installer does some IDE stuff which registers the .mgcb files with the Pipeline tool. that won't work in a nuget I don't think. Also the Pipeline Nuget won't work on a Mac. 

Basically this is a good idea but there are a bunch of things we need to think about, I really want the MG experience for the end user to be straight forward with as little confusion as possible which is why we worked really hard to make sure the Mac Content Pipeline stuff works just as it does on windows. The project is now at a point where we need to polish the user experience and make things nice and a pleasure to use :)

**Comment by DDReaper** [comment]
Well, you know my feelings @dellis1972 , I'd prefer one delivery solution, which dramatically eases the adoption.  With WindowsUniversal and .NET going all in with NuGet it would make sense.
I also like ProtoBuild's packaging process, although that has wider implications within the project.

If we could cache the NuGet's with the templates, it would ease the offline process but alas that is against NuGet's core principles.  

If we could find a way to ship the packages NuGet's with the build whilst still retaining the reference to the online versions, that would be an ideal solution.  Alternatively, get buy in from the wider audience for ProtoBuild packages as they deploy to all of our target platforms.

Choices Choices

**Comment by hach-que** [comment]
This might be relevant, with #4295 you can just create an `AdditionalProjectTransforms.xslt` for the NuGet project overlay.  This XSLT file would contain the additional MSBuild XML that's required for each platform, and obviously because it's XSLT you can switch out what project XML gets written out based on the host platform.  This XSLT is used when generating the projects, but all of the Protobuild-related assets get removed when generating platforms with Standard C# Projects (instead of Protobuild).

Thus with #4295, creating projects through either the Visual Studio or Xamarin Studio project creation wizards you'll have everything wired up correctly for any project types.  So it might implicitly solve the props / targets problem by just doing it at that level instead.

**Comment by tgjones** [comment]
> If we could cache the NuGet's with the templates, it would ease the offline process but alas that is against NuGet's core principles.

NuGet has supported exactly this feature - pre-installed, offlined, cached NuGet packages in VS templates - for a while. (That's how ASP.NET MVC templates work.)
https://docs.nuget.org/create/packages-in-visual-studio-templates

**Comment by dellis1972** [comment]
We also do this with the Xamarin Studio Addin, we cache the packages
locally at install time.

**Comment by harry-cpp** [comment]
Templates are using nugets now, and nugets do set the required props automatically so closing this.
